clang/test/CIR/CodeGen/union.c

```diff
@@ -5,25 +5,146 @@
+// LLVM: %union.U1 = type { i32 }
+// OGCG: %union.U1 = type { i32 }
+
+union U2 {
+  char b;
+  short s;
+  int i;
+  float f;
+  double d;
+};
+
+// CIR:  !rec_U2 = !cir.record<union "U2" {!s8i, !s16i, !s32i, !cir.float, !cir.double}>
+// LLVM: %union.U2 = type { double }
+// OGCG: %union.U2 = type { double }
+
 union IncompleteU *p;
 
-// CIR: cir.global external @p = #cir.ptr<null> : !cir.ptr<!rec_IncompleteU>
+// CIR:  cir.global external @p = #cir.ptr<null> : !cir.ptr<!rec_IncompleteU>
 // LLVM: @p = dso_local global ptr null
 // OGCG: @p = global ptr null, align 8
 
-void f(void) {
+void f1(void) {
   union IncompleteU *p;
 }
 
-// CIR: cir.func @f()
-// CIR-NEXT: cir.alloca !cir.ptr<!rec_IncompleteU>, !cir.ptr<!cir.ptr<!rec_IncompleteU>>, ["p"]
-// CIR-NEXT: cir.return
+// CIR:      cir.func @f1()
+// CIR-NEXT:   cir.alloca !cir.ptr<!rec_IncompleteU>, !cir.ptr<!cir.ptr<!rec_IncompleteU>>, ["p"]
+// CIR-NEXT:   cir.return
 
-// LLVM:      define void @f()
+// LLVM:      define void @f1()
 // LLVM-NEXT:   %[[P:.*]] = alloca ptr, i64 1, align 8
 // LLVM-NEXT:   ret void
 
-// OGCG:      define{{.*}} void @f()
+// OGCG:      define{{.*}} void @f1()
 // OGCG-NEXT: entry:
 // OGCG-NEXT:   %[[P:.*]] = alloca ptr, align 8
 // OGCG-NEXT:   ret void
+
+int f2(void) {
+  union U1 u;
+  u.n = 42;
+  return u.n;
+}
+
+// CIR:      cir.func @f2() -> !s32i
+// CIR-NEXT:   %0 = cir.alloca !s32i, !cir.ptr<!s32i>, ["__retval"] {alignment = 4 : i64}
+// CIR-NEXT:   %1 = cir.alloca !rec_U1, !cir.ptr<!rec_U1>, ["u"] {alignment = 4 : i64}
+// CIR-NEXT:   %2 = cir.const #cir.int<42> : !s32i
+// CIR-NEXT:   %3 = cir.get_member %1[0] {name = "n"} : !cir.ptr<!rec_U1> -> !cir.ptr<!s32i>
+// CIR-NEXT:   cir.store %2, %3 : !s32i, !cir.ptr<!s32i>
+// CIR-NEXT:   %4 = cir.get_member %1[0] {name = "n"} : !cir.ptr<!rec_U1> -> !cir.ptr<!s32i>
+// CIR-NEXT:   %5 = cir.load %4 : !cir.ptr<!s32i>, !s32i
+// CIR-NEXT:   cir.store %5, %0 : !s32i, !cir.ptr<!s32i>
+// CIR-NEXT:   %6 = cir.load %0 : !cir.ptr<!s32i>, !s32i
+// CIR-NEXT:   cir.return %6 : !s32i
+
+// LLVM:      define i32 @f2()
+// LLVM-NEXT:   %1 = alloca i32, i64 1, align 4
+// LLVM-NEXT:   %2 = alloca %union.U1, i64 1, align 4
+// LLVM-NEXT:   store i32 42, ptr %2, align 4
+// LLVM-NEXT:   %3 = load i32, ptr %2, align 4
+// LLVM-NEXT:   store i32 %3, ptr %1, align 4
+// LLVM-NEXT:   %4 = load i32, ptr %1, align 4
+// LLVM-NEXT:   ret i32 %4
+
+//      OGCG: define dso_local i32 @f2()
+// OGCG-NEXT: entry:
+// OGCG-NEXT: %u = alloca %union.U1, align 4
+// OGCG-NEXT: store i32 42, ptr %u, align 4
+// OGCG-NEXT: %0 = load i32, ptr %u, align 4
+// OGCG-NEXT: ret i32 %0
+
+
+void shouldGenerateUnionAccess(union U2 u) {
+  u.b = 0;
+  u.b;
+  u.i = 1;
+  u.i;
+  u.f = 0.1F;
+  u.f;
+  u.d = 0.1;
+  u.d;
+}
+
+// CIR:      cir.func @shouldGenerateUnionAccess(%arg0: !rec_U2
+// CIR-NEXT:   %0 = cir.alloca !rec_U2, !cir.ptr<!rec_U2>, ["u", init] {alignment = 8 : i64}
```

Comment:
SSA values in tests should use FileCheck regexes and never the output value directly. Please change this for all CIR, LLVM, etc.